File: src/converter/plugins/convertTypeOperator.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";

export const convertTypeOperator = createSimplePlugin((node, context) => {
    if (!ts.isTypeOperatorNode(node)) return null

    if (
        node.operator === ts.SyntaxKind.UniqueKeyword
        && node.type.kind === ts.SyntaxKind.SymbolKeyword
    ) {
        const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
        checkCoverageService?.cover(node)
        checkCoverageService?.cover(node.type)

        return "/* unique */ js.symbol.Symbol"
    }

    return null
})

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertTypeOperator.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createSimplePlugin
import typescript.Node
import typescript.SyntaxKind
import typescript.isTypeOperatorNode

val convertTypeOperator = createSimplePlugin plugin@{ node: Node, context, _ ->
    if (!isTypeOperatorNode(node)) return@plugin null

    if (
        node.operator == SyntaxKind.UniqueKeyword
        && node.type.kind == SyntaxKind.SymbolKeyword
    ) {
        val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
        checkCoverageService?.cover(node)
        checkCoverageService?.cover(node.type)

        return@plugin "/* unique */ js.symbol.Symbol"
    }

    null
}
